### PR TITLE
[DAGCombiner] Remove a hasOneUse check in visitAND

### PR DESCRIPTION
**Pull request by david-arm**
For some reason there was a hasOneUse check on the splat for the
second operand and it's not obvious to me why. The check blocks
optimisations for lowering of nodes like AVGFLOORU and AVGCEILU.

In a follow-on patch I also plan to improve the generated code
for AVGCEILU further by teaching computeKnownBits about
zero-extending masked loads.

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/avg.ll [hunk 1]
```diff
@@ -17,6 +17,28 @@ define <16 x i16> @zext_avgflooru(<16 x i8> %a0, <16 x i8> %a1) {
   ret <16 x i16> %avg
 }
 
+define void @zext_mload_avgflooru(ptr %p1, ptr %p2, <vscale x 8 x i1> %mask) {
```
Comment (davemgreen) [inline llvm/test/CodeGen/AArch64/avg.ll:20]: This feels like it should be in a sve file, like sve-hadd.ll

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 2]
```diff
@@ -7095,8 +7095,7 @@ SDValue DAGCombiner::visitAND(SDNode *N) {
     // fold (and (masked_load) (splat_vec (x, ...))) to zext_masked_load
     auto *MLoad = dyn_cast<MaskedLoadSDNode>(N0);
     ConstantSDNode *Splat = isConstOrConstSplat(N1, true, true);
-    if (MLoad && MLoad->getExtensionType() == ISD::EXTLOAD && Splat &&
-        N1.hasOneUse()) {
```
Comment (davemgreen) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:7099]: Should it be checking N0.hasOneUse()? (i.e the load)

Comment (david-arm) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:7099]: If I add that check it causes regression in CodeGen/AArch64/sve-load-compare-store.ll and CodeGen/Thumb2/mve-masked-load.ll

Comment (david-arm) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:7099]: An extra instruction in multi_user_zext (mve-masked-load.ll) and two extra in sve_load_compare_store (sve-load-compare-store.ll)

Comment (paulwalker-arm) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:7099]: You can see from the code the transformation has explicit handling for this.  The original load is an any-extend and so the transformation is simply promoting this to a zero-extend which can be used by all uses.

File: llvm/test/CodeGen/AArch64/avg.ll (hunk 1 above)
Comment (david-arm) [reply llvm/test/CodeGen/AArch64/avg.ll]: Done!

**Review by paulwalker-arm — APPROVED** [review]
FYI: Not affected by this PR but I think the original transformation might have a bug because when promoting the any-extend to zero-extend, I think passthru should also be explicitly zero-extended.

I'm not 100% sure because an equally valid interpretation would be for users of passthru to apply the extension type.

My guess is with most all uses being either undef or zero this isn't actually affects us yet.

**Comment by david-arm** [comment]
I ran make check-all downstream on Linux and it passed. Not sure what is happening with the pre-commit testing, but the Windows build passed fine.

**Pull request merged**